src/Marten/Events/Daemon/ProjectionAgent.cs

```diff
@@ -0,0 +1,191 @@
+            _logger = logger;
+
+            var singleFile = new ExecutionDataflowBlockOptions
+            {
+                EnsureOrdered = true,
+                MaxDegreeOfParallelism = 1,
+                CancellationToken = _cancellationSource.Token
+            };
+
+            _commandBlock = new ActionBlock<Command>(processCommand, singleFile);
+
+            _controller =
+                new ProjectionController(projectionShard.ProjectionOrShardName, this, projectionShard.Options);
+
+            _loader = new TransformBlock<EventRange, EventRange>(loadEvents, singleFile);
+        }
+
+        // TODO -- use IAsyncDisposable
+        public async Task Stop()
+        {
+            _cancellationSource.Cancel();
+            _commandBlock.Complete();
+            await _commandBlock.Completion;
+
+            _loader.Complete();
+            await _loader.Completion;
+
+            _hopper.Complete();
+            await _hopper.Completion;
+
+            await _projectionShard.Stop();
+
+            _subscription.Dispose();
+
+        }
+
+        private async Task<EventRange> loadEvents(EventRange range)
+        {
+            try
+            {
+                // TODO -- resiliency here.
+                await _fetcher.Load(range, _cancellationSource.Token);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Error loading events for " + range);
+                // TODO -- retry? circuit breaker?
+            }
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                _logger.LogDebug($"Loaded events for {range}");
+            }
+
+            return range;
+        }
+
+        private void processCommand(Command command) => command.Apply(_controller);
+
+        public AgentStatus Status { get; private set; }
+
+
+        public void StartRange(EventRange range)
+        {
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                _logger.LogDebug("Enqueued processing of " + range);
+            }
+
+            _loader.Post(range);
+        }
+
+        public async Task Start(ShardStateTracker tracker)
+        {
+            _logger.LogInformation($"Starting projection agent for '{_projectionShard.ProjectionOrShardName}'");
+
+            _tracker = tracker;
+
+
+            _fetcher = new EventFetcher(_store, _projectionShard.EventFilters);
+            _hopper = _projectionShard.Start(this, _logger, _cancellationSource.Token);
+            _loader.LinkTo(_hopper);
+
+            var lastCommitted = await _store.Events.ProjectionProgressFor(_projectionShard.ProjectionOrShardName);
+
+            _commandBlock.Post(Command.Started(tracker.HighWaterMark, lastCommitted));
+
+            _subscription = _tracker.Subscribe(this);
+
+            _logger.LogInformation($"Projection agent for '{_projectionShard.ProjectionOrShardName}' has started from sequence {lastCommitted} and a high water mark of {tracker.HighWaterMark}");
```

Comment:
I'd be much appreciated if you used the structured logging overloads for the `ILogger` integration. It's useful for those of us using structured sinks and log aggregation platforms like Sumologic.

```suggestion
            _logger.LogInformation("Projection agent for '{ProjectionOrShardName}' has started from sequence {lastCommitted} and a high water mark of {HighWaterMark}", _projectionShard.ProjectionOrShardName, lastCommitted, tracker.HighWaterMark);
```

Reply:
@Hawxy I think that's a great idea, and had thought vaguely about that. Any chance you'd be interested in taking that on?

Reply:
Sure 👍